### ISSUE TITLE
 When to use Deployment + PVC vs Statefulset with volumeClaimTemplates

### ISSUE THREAD

**Issue by xMAC94x** [bitnami/charts#8138]
**Which chart**:
Wikimedia (and others)

**Is your feature request related to a problem? Please describe.**
The behavior difference of Statefulsets and Deployments regarding PVC

**Describe the solution you'd like**
More of a question i want to introduce:
Why does the Mediawiki Helm chart use a Deployment + PVC https://github.com/bitnami/charts/blob/master/bitnami/mediawiki/templates/deployment.yaml#L3
while the Postgresql Helm chart is using a Statefulset with volumeClaimTemplates configures https://github.com/bitnami/charts/blob/master/bitnami/postgresql/templates/statefulset.yaml#L2 ?

Okay lets go a bit deeper into the question, but first let me tell you that i am kind of an k8s/helm noob. In my naive though i would use a deployment for something that has no state, and a Statefulset ALWAYS when something has state. A Database has state, sure. But the mediawiki above also has state, hasn’t it ? Sure it stores it users and posts in the respective Database (MariaDB/Postgresql) but it also needs a PVC for

    Its configuration (which could a k8s config though)
    Its themes (which prob could be a normal volume)
    its uploaded images (which is actually a valid point for the PVC imo)

So i am wondering why it was decided to use a statefulset for the postgresdb but dont go with statefulset for mediawiki.

On the first glance one could prob think there is no difference between a deployment with a pvc and a statefulset, but this is not exactly true. a Statefulset which has a PVC created by its volumeClaimTemplates will keep the PVC existing even if the helm chart gets deleted. And i am a very clumsy person that would like to move my homelab to kubernetes but would minimise the risk of accidentally deleting my whole persistence.

So actually i hope that the result is that mediawiki in fact should be a Statefulset, and if this turns out to be the case i would be happy to provide a PR :wink:

**Comment by marcosbc** [comment]
Hi @xMAC94x, actually using deployments is the appropriate use case here. That is because the end user is expected to access only one instance of the application (i.e. you would not want to access different wikis each time you reload due to a load balancer), so there is only one "state" of the MediaWiki application.

However, in the case of databases, this is another scenario because, in the case of MariaDB/MySQL, each node is expected to act as a primary/replica, so each node has one unique state.

To summarize, supposing an application is configured to use N volumes:

* Deployments: 1 state, for all replicas combined - There is only N volumes in total (and when replicas > 1, [ReadWriteMany must be configured](https://github.com/bitnami/charts/tree/master/bitnami/mediawiki#prerequisites) so the volumes are writable in all Kubernetes nodes)
* Statefulset: 1 state for each replica - Each node have N unique volumes, there are N * replicas volumes in total

**Comment by xMAC94x** [comment]
Hi @marcosbc  thanks for the explanation.
Ah i see, it makes sense when you use replication and have multiple instances. I haven't thought about it.

Regarding the problem of PVC deletion (in case someone googles this) I found out that you can annotate your ressources with
```yaml
annotations:
        "helm.sh/resource-policy": keep
```
and then helm 3 does keep them after uninstall of the chart. (make sure that the annotation is present at chart creation time e.g. by setting the right annotation in your values.yaml).

Closing this ticket now :)
